src/it/scala/io/iohk/ethereum/transactionTest/util/DumpChainActor.scala

```diff
@@ -0,0 +1,174 @@
+        actor ! SendMessage(GetNodeData(mptRoots))
+        stateNodesHashes = stateNodesHashes ++ mptRoots.toSet
+      }
+
+    case MessageReceived(m: BlockBodies) =>
+      m.bodies.zip(blockHeadersHashes).foreach { case (b, h) =>
+        blockBodyStorage = blockBodyStorage + (h -> b)
+      }
+      val bodiesFile = new FileWriter("bodies.txt", true)
+      blockBodyStorage.foreach{case (h,v) =>  bodiesFile.write(s"${Hex.toHexString(h.toArray[Byte])} ${Hex.toHexString(encode(v))}\n")}
+      bodiesFile.close()
+
+    case MessageReceived(m: Receipts) =>
+      m.receiptsForBlocks.zip(blockHeadersHashes.tail).foreach { case (r, h) =>
+        blockReceiptsStorage = blockReceiptsStorage + (h -> r)
+      }
+      val receiptsFile = new FileWriter("receipts.txt", true)
+      blockReceiptsStorage.foreach { case (h, v: Seq[Receipt]) => receiptsFile.write(s"${Hex.toHexString(h.toArray[Byte])} ${Hex.toHexString(encode(toRlpList(v)))}\n") }
+      receiptsFile.close()
+
+    case MessageReceived(m: NodeData) =>
+
+      val stateNodes = m.values.filter(node => stateNodesHashes.contains(kec256(node)))
+      val contractNodes = m.values.filter(node => contractNodesHashes.contains(kec256(node)))
+      val evmCode = m.values.filter(node => evmCodeHashes.contains(kec256(node)))
+
+      val nodes = NodeData(stateNodes).values.indices.map(i => NodeData(stateNodes).getMptNode(i))
+
+      val children = nodes.flatMap {
+        case n: MptBranch => n.children.collect { case Left(h: MptHash) if h.hash.nonEmpty => h.hash }
+        case MptExtension(_, Left(h)) => Seq(h.hash)
+        case n: MptLeaf => Seq.empty
+        case _ => Seq.empty
+      }
+
+      var cChildren: Seq[ByteString] = Nil
+      var evmTorequest: Seq[ByteString] = Nil
+
+      nodes.foreach {
+        case n: MptLeaf =>
+          if (n.getAccount.codeHash != DumpChainActor.emptyEvm) {
+            peers.headOption.foreach { case Peer(_, actor) =>
+              evmTorequest = evmTorequest :+ n.getAccount.codeHash
+              evmCodeHashes = evmCodeHashes + n.getAccount.codeHash
+            }
+          }
+          if (n.getAccount.storageRoot != DumpChainActor.emptyStorage) {
+            peers.headOption.foreach { case Peer(_, actor) =>
+              cChildren = cChildren :+ n.getAccount.storageRoot
+              contractNodesHashes = contractNodesHashes + n.getAccount.storageRoot
+            }
+          }
+        case _ =>
+      }
+
+      val cNodes = NodeData(contractNodes).values.indices.map(i => NodeData(contractNodes).getMptNode(i))
+      cChildren = cChildren ++ cNodes.flatMap {
+        case n: MptBranch => n.children.collect { case Left(h: MptHash) if h.hash.nonEmpty => h.hash }
+        case MptExtension(_, Left(h)) => Seq(h.hash)
+        case _ => Seq.empty
+      }
+      peers.headOption.foreach { case Peer(_, actor) =>
+        actor ! SendMessage(GetNodeData(children ++ cChildren ++ evmTorequest))
```

Comment:
Minor detail: Should we check whether there is any node to ask for before asking for it? I tried using our dumper with Geth (which seems to be working :smile:) and it got in an infinite request for no data with `GetNodeData`